lib/Sema/TypeCheckType.cpp

```diff
@@ -910,10 +910,38 @@ static Type resolveTypeDecl(TypeDecl *typeDecl, SourceLoc loc,
     return ErrorType::get(ctx);
   }
 
-  if (type->hasError() && isa<AssociatedTypeDecl>(typeDecl)) {
-    maybeDiagnoseBadConformanceRef(fromDC,
-                                   foundDC->getDeclaredInterfaceType(),
-                                   loc, cast<AssociatedTypeDecl>(typeDecl));
+  if (type->hasError()) {
+    if (auto associatedTypeDecl = dyn_cast<AssociatedTypeDecl>(typeDecl)) {
+      maybeDiagnoseBadConformanceRef(fromDC,
+                                     foundDC->getDeclaredInterfaceType(),
+                                     loc, associatedTypeDecl);
+    }
+    if (auto aliasDecl = dyn_cast<TypeAliasDecl>(typeDecl)) {
```

Comment:
Perhaps we can call maybeDiagnoseBadConformanceRef() unconditionally, if its a typealias in a protocol. It's a hack anyway, so limiting it to direct associated type references isn't necessary.